File: Dorkbox-Util/src/dorkbox/util/properties/SortedProperties.java
/*
 * Copyright 2010 dorkbox, llc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dorkbox.util.properties;

import java.util.Collections;
import java.util.Comparator;
import java.util.Enumeration;
import java.util.Properties;
import java.util.Vector;

public class SortedProperties extends Properties {

    private static final long serialVersionUID = 3988064683926999433L;

    private final Comparator<Object> compare = new Comparator<Object>() {
        @Override
        public int compare(Object o1, Object o2) {
            return o1.toString().compareTo(o2.toString());
        }};

    @Override
    public synchronized Enumeration<Object> keys() {
        Enumeration<Object> keysEnum = super.keys();

        Vector<Object> vector = new Vector<Object>(size());
        for (;keysEnum.hasMoreElements();) {
            vector.add(keysEnum.nextElement());
        }

        Collections.sort(vector, this.compare);

        return vector.elements();
    }
}

File: src/dorkbox/util/properties/SortedProperties.kt
/*
 * Copyright 2023 dorkbox, llc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dorkbox.util.properties

import java.util.*

class SortedProperties : Properties() {
    private val compare = Comparator<Any> { o1, o2 ->
        o1.toString().compareTo(o2.toString())
    }

    @Synchronized
    override fun keys(): Enumeration<Any> {
        val keysEnum = super.keys()
        val vector: Vector<Any> = Vector<Any>(this.size)

        while (keysEnum.hasMoreElements()) {
            vector.add(keysEnum.nextElement())
        }

        Collections.sort(vector, compare)
        return vector.elements()
    }

    companion object {
        private const val serialVersionUID = 3988064683926999433L

        /**
         * Gets the version number.
         */
        val version = "1.42"
    }
}
